Add GitHub Actions guide (#11619)
docker/docker.github.io: ci-cd/github-actions.md

title: Configure GitHub Actions
---

This page guides you through the process of setting up a GitHub Action CI/CD pipeline with Docker containers. Before setting up a new pipeline, we recommend that you take a look at [Ben's blog](https://www.docker.com/blog/best-practices-for-using-docker-hub-for-ci-cd/){:target="_blank" rel="noopener" class="_"} on CI/CD best practices .

This guide contains instructions on how to:

1. Use a sample Docker project as an example to configure GitHub Actions
2. Set up the GitHub Actions workflow
3. Optimize your workflow to reduce the number of pull requests and the total build time, and finally,
4. Push only specific versions to Docker Hub.

## Set up a Docker project

Let’s get started. This guide uses a simple Docker project as an example. The [SimpleWhaleDemo](https://github.com/usha-mandya/SimpleWhaleDemo){:target="_blank" rel="noopener" class="_"} repository contains an Ngnix alpine image. You can either clone this repository, or use your own Docker project.

![SimpleWhaleDemo](images/simplewhaledemo.png){:width="500px"}

Before we start, ensure you can access [Docker Hub](https://hub.docker.com/) from any workflows you create. To do this:

> 1. Add your Docker ID and Personal Access Token (PAT) as secrets to GitHub.

Reviewer: ```suggestion
1. Add your Docker ID as a secret to GitHub.
```
Navigate to your GitHub repository and click **Settings > Secrets**.

![github-new-secret](https://user-images.githubusercontent.com/207759/97191354-17ba6180-17a7-11eb-9f4c-6586a02bdb91.jpg)

Create a new secret with the name `DOCKER_HUB_USERNAME` and with your Docker ID as value.

2. To create a new token, go to [Docker Hub Settings](https://hub.docker.com/settings/security) and then click **New Access Token**.
3. Let’s call this token **simplewhaleci**.

    ![New access token](images/github-access-token.png){:width="500px"}

> 4. Now, add this token to your username as secrets into the GitHub secrets UI.

Reviewer: ```suggestion
4. Now, add this Personal Access Token (PAT) as a second secret into the GitHub secrets UI with the name `DOCKER_HUB_ACCESS_TOKEN`.
```

Reviewer: I suggest to write the secret names in the docs to make it readable in addition to the screenshots.

2. The second one is the build and push action.

In this example, let us set the push flag to `true` as we also want to push. We’ll then add a tag to specify to always go to the latest version. Lastly, we’ll echo the image digest to see what was pushed.

To set up the workflow:

1. Go to your repository in GitHub and then click **Actions** > **New workflow**.
2. Click **set up a workflow yourself** and add the following content:

First, we will name this workflow:

```yaml
name: CI to Docker Hub
 ```

Then, we will choose when we run this workflow. In our example, we are going to do it for every push against the main branch of our project:

```yaml
on:
  push:
    branches: [ main ]
```

Now, we need to specify what we actually want to happen within our action (what jobs), we are going to add our build one and select that it runs on the latest Ubuntu instances available:

```yaml
jobs:

  build:
    runs-on: ubuntu-latest
```

Now, we can add the steps required. The first one is to use our PAT and username to log into Docker Hub. The second is the Builder, the action  uses BuildKit under the hood through a simple Buildx action which we will also setup

```yaml

>     Steps:

Reviewer: ```suggestion
    steps:
```

lower case

Author: Fixed

      - name: Login to Docker Hub
        uses: docker/login-action@v1
        with:

>           username: ${{ secrets.DOCKER_HUB_USERNAME }}

Reviewer: Oh, these double curly braces here and further down in the code samples get lost in Jekyll. 😢 

![docs-yaml-vars-missing](https://user-images.githubusercontent.com/207759/97193300-2570e680-17a9-11eb-9523-e1c54ca6faa9.jpg)

Looks like there are some workarounds to escape the curly braces https://stackoverflow.com/questions/24102498/escaping-double-curly-braces-inside-a-markdown-code-block-in-jekyll

Author: I've added `{% raw %}` and `{% endraw %}` tags

Author: ![image](https://user-images.githubusercontent.com/47779042/97238697-ead77000-17e1-11eb-85eb-6d75f6b2f18b.png)

        with:
          context: ./
          file: ./Dockerfile
          push: true
          tags: ushamandya/simplewhale:latest

      - name: Image digest
        run: echo ${{ steps.docker_build.outputs.digest }}
```

Now, let the workflow run for the first time and then tweak the Dockerfile to make sure the CI is running and pushing the new image changes:

![CI to Docker Hub](images/ci-to-hub.png){:width="500px"}

## Optimizing the workflow

Next, let’s look at how we can optimize the GitHub Actions workflow through build cache. This has two main advantages:

1. Build cache reduces the build time as it will not have to re-download all of the images, and
2. It also reduces the number of pulls we complete against Docker Hub. We need to make use of GitHub cache to make use of this.

Let us set up a Builder with a build cache. First, we need to set up cache for the builder.  In this example, let us add the path and keys to store this under using GitHub cache for this.

```yaml

      - name: Cache Docker layers
        uses: actions/cache@v2
        with:
          path: /tmp/.buildx-cache
          key: ${{ runner.os }}-buildx-${{ github.sha }}
          restore-keys: |

>             ${{ runner.os }}-buildx-

Reviewer: The `restore-keys` is different to `key`?
I just want to double check if this is correct. 😅 

Seems correct as it is used in https://github.com/metcalfc/docker-action-examples/blob/main/.github/workflows/release.yml#L44-L50 👍

        uses: docker/login-action@v1
        with:
        registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.GHCR_TOKEN }}
```

Remember to change how the image is tagged. The following example keeps ‘latest’ as the only tag. However, you can add any logic to this if you prefer:

```yaml
  tags: ghcr.io/${{ github.repository_owner }}/simplewhale:latest
```

![Update tagged images](images/ghcr-logic.png){:width="500px"}

Now, we will have two different flows: one for our changes to master, and one for our pull requests. Next, we need to modify what we had before to ensure we are pushing our PRs to the GitHub registry rather than to Docker Hub.

## Conclusion

In this guide, you have learnt how to set up GitHub Actions workflow to an existing Docker project, optimize your workflow to improve build times and reduce the number of pull requests, and finally, we learnt how to push only specific versions to Docker Hub.

## Next steps

You can now consider setting up nightly tests against the latest tag, test each PR, or do something more elegant with the tags we are using and make use of the Git tag for the same tag in our image.

> To look at how you can do one of these, or to get a full example on how to set up what we have accomplished today, check out Chad’s repo which runs you through this and more details on our latest GitHub action - https://github.com/metcalfc/docker-action-examples.

Reviewer: Do we want to make it a link?

```suggestion
To look at how you can do one of these, or to get a full example on how to set up what we have accomplished today, check out Chad’s repo which runs you through this and more details on our latest GitHub action - [https://github.com/metcalfc/docker-action-examples](https://github.com/metcalfc/docker-action-examples).
```

Author: Updated

      - name: Set up Docker Buildx
        id: buildx
        uses: docker/setup-buildx-action@v1

      - name: Build and push
        id: docker_build
        uses: docker/build-push-action@v2
        with:
          context: ./
          file: ./Dockerfile
          push: true
          tags: ushamandya/simplewhale:latest

      - name: Image digest
        run: echo ${{ steps.docker_build.outputs.digest }}
```
{% endraw %}

Now, let the workflow run for the first time and then tweak the Dockerfile to make sure the CI is running and pushing the new image changes:

![CI to Docker Hub](images/ci-to-hub.png){:width="500px"}

## Optimizing the workflow

Next, let’s look at how we can optimize the GitHub Actions workflow through build cache. This has two main advantages:

1. Build cache reduces the build time as it will not have to re-download all of the images, and
2. It also reduces the number of pulls we complete against Docker Hub. We need to make use of GitHub cache to make use of this.

Let us set up a Builder with a build cache. First, we need to set up cache for the builder.  In this example, let us add the path and keys to store this under using GitHub cache for this.

> ```yaml

Reviewer: This needs another `{% raw %} ... {% endraw %}` block.

This involves two steps:

1. Modifying the GitHub workflow to only push commits with specific tags to Docker Hub
2. Setting up a GitHub Actions file to store the latest commit as an image in the GitHub registry

First, let us modify our existing GitHub workflow to only push to Hub if there’s a particular tag. For example:

```yaml
on:
  push:
    tags:
      - "v*.*.*"
```

This ensures that the main CI will only trigger if we tag our commits with `V.n.n.n.` Let’s test this. For example, run the following command:

```bash
git tag -a v1.0.2
git push origin v1.0.2
```

Now, go to GitHub and check your Actions

![Push tagged version](images/push-tagged-version.png){:width="500px"}

Now, let’s set up a second GitHub action file to store our latest commit as an image in the GitHub registry. You may want to do this to:

1. Run your nightly tests or recurring tests, or
2. To share work in progress images with colleagues.

Let’s clone our previous GitHub action and add back in our previous logic for all pushes. This will mean we have two workflow files, our previous one and our new one we will now work on.
Next, change your Docker Hub login to a GitHub container registry login:

> ```yaml

Reviewer: This needs another `{% raw %} ... {% endraw %}` block.

        if: github.event_name != 'pull_request'
        uses: docker/login-action@v1
        with:
        registry: ghcr.io
          username: ${{ github.repository_owner }}
          password: ${{ secrets.GHCR_TOKEN }}
```

Remember to change how the image is tagged. The following example keeps ‘latest’ as the only tag. However, you can add any logic to this if you prefer:

> ```yaml

Reviewer: This needs another `{% raw %} ... {% endraw %}` block.
